backend/msgpack/topology.go

```diff
@@ -0,0 +1,175 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"sync"
+
+	schema "github.com/m3db/m3cluster/generated/proto/placement"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3collector/runtime"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3x/clock"
+)
+
+var (
+	errNilValue                  = errors.New("nil value received")
+	errTopologyIsOpenOrClosed    = errors.New("topology is already open or closed")
+	errTopologyIsNotOpenOrClosed = errors.New("topology is not open or closed")
+	errNoEligiblePlacement       = errors.New("no eligible placement")
+)
+
+type topologyState int
+
+const (
+	topologyNotOpen topologyState = iota
+	topologyOpen
+	topologyClosed
+)
+
+// topology stores topology information.
+type topology interface {
+	// Open opens the topology.
+	Open() error
+
+	// Route routes a metric alongside its policies.
+	Route(mu unaggregated.MetricUnion, vp policy.VersionedPolicies) error
+
+	// Close closes the topology.
+	Close() error
+}
+
+type topo struct {
+	sync.RWMutex
+	runtime.Value
+
+	serverOpts    ServerOptions
+	placementOpts placementOptions
+	nowFn         clock.NowFn
+
+	state     topologyState
+	proto     *schema.PlacementSnapshots
+	placement placement
+}
+
+// newTopology creates a new topology.
+func newTopology(writerMgr instanceWriterManager, opts ServerOptions) topology {
+	clockOpts := opts.ClockOptions()
+	topology := &topo{
+		serverOpts: opts,
+		nowFn:      clockOpts.NowFn(),
+		proto:      &schema.PlacementSnapshots{},
+	}
+	topologyOpts := opts.TopologyOptions()
+	topology.placementOpts = newPlacementOptions().
+		SetClockOptions(clockOpts).
+		SetHashGenFn(topologyOpts.HashGenFn()).
+		SetInstanceWriterManager(writerMgr)
+	valueOpts := runtime.NewOptions().
+		SetInstrumentOptions(opts.InstrumentOptions()).
+		SetInitWatchTimeout(topologyOpts.InitWatchTimeout()).
+		SetKVStore(topologyOpts.KVStore()).
+		SetUnmarshalFn(topology.toPlacement).
+		SetProcessFn(topology.process)
+	topology.Value = runtime.NewValue(topologyOpts.TopologyKey(), valueOpts)
+	return topology
+}
+
+func (t *topo) Open() error {
+	t.Lock()
+	if t.state != topologyNotOpen {
+		return errTopologyIsOpenOrClosed
+	}
+	t.state = topologyOpen
+	t.Unlock()
+
+	// NB(xichen): we watch the topology updates outside the lock because
+	// otherwise the initial update will trigger the process() callback,
+	// which attempts to acquire the same lock, causing a deadlock.
+	return t.Watch()
+}
+
+func (t *topo) Route(mu unaggregated.MetricUnion, vp policy.VersionedPolicies) error {
+	t.RLock()
```

Comment:
Nit: Wonder if we will have lots of time spent acquiring read locks - it seems as if each write will require at least 4 read locks. Would be worth looking at profiles once we get some load tests going.

Alternate design may be to acquire some sort of lock on the higher level (topology) per snapshot call? This would mean we cannot process a change mid snapshot, but that may be OK? Or maybe we can get a lock that allows multiple writes that we track via ref counts or something. Something to keep in mind if we need to optimize later.